File: app/src/main/java/net/sistransito/mobile/main/InstallTask.java
/**
 * Load sqlite chipper library
 * initialized all database object
 * initialized all single tone object
 */

package net.sistransito.mobile.main;

import android.os.AsyncTask;

import net.sistransito.MainActivity;
import net.sistransito.Application;
import net.sistransito.mobile.appobject.AppObject;
import net.sqlcipher.database.SQLiteDatabase;

public class InstallTask extends AsyncTask<Void, Integer, Void> {

    private MainActivity mActivity;

    public InstallTask(MainActivity main) {
        attach(main);
    }

    void attach(MainActivity activity) {
        this.mActivity = activity;

    }

    @Override
    protected Void doInBackground(Void... arg0) {

        publishProgress(0);

        try {
            SQLiteDatabase.loadLibs(mActivity);
            publishProgress(80);
            AppObject.newInstance(Application.getAppContext());
        } catch (Exception e) {

        }

        publishProgress(100);
        return null;
    }

    @Override
    protected void onProgressUpdate(Integer... progress) {
        if (mActivity == null) {
        } else {
            mActivity.updateProgress(progress[0]);
        }
    }

    @Override
    protected void onPostExecute(Void result) {
        super.onPostExecute(result);
        if (mActivity == null) {
        } else {
            mActivity.installWorkComplete();
        }

    }

    void detach() {
        mActivity = null;
    }
}

File: app/src/main/java/net/sistransito/mobile/main/InstallTask.kt
package net.sistransito.mobile.main

import android.content.Context
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import net.sistransito.mobile.appobject.AppObject
import net.sqlcipher.database.SQLiteDatabase

class InstallTask(
    private val activity: LifecycleOwner,
    private val context: Context,
    private val updateProgress: (Float) -> Unit,
    private val onComplete: () -> Unit
) {

    init {
        activity.lifecycleScope.launch {
            activity.repeatOnLifecycle(Lifecycle.State.STARTED) {
                executeTask()
            }
        }
    }

    private suspend fun executeTask() {
        updateProgress(0f)

        try {
            withContext(Dispatchers.IO) {
                SQLiteDatabase.loadLibs(context)
            }
            updateProgress(80f)
            AppObject.newInstance(context.applicationContext)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        updateProgress(100f)
        onComplete()
    }
}
